src/storage/disk_table.cc

```diff
@@ -850,6 +850,237 @@ void DiskTableTraverseIterator::NextPK() {
+      it_(it),
+      snapshot_(snapshot),
+      ttl_type_(ttl_type),
+      expire_time_(expire_time),
+      expire_cnt_(expire_cnt),
+      has_ts_idx_(false),
+      ts_idx_(0) {}
+
+DiskTableKeyIterator::DiskTableKeyIterator(rocksdb::DB* db, rocksdb::Iterator* it, const rocksdb::Snapshot* snapshot,
+                                           ::openmldb::storage::TTLType ttl_type, const uint64_t& expire_time,
+                                           const uint64_t& expire_cnt, int32_t ts_idx)
+    : db_(db),
+      it_(it),
+      snapshot_(snapshot),
+      ttl_type_(ttl_type),
+      expire_time_(expire_time),
+      expire_cnt_(expire_cnt),
+      has_ts_idx_(true),
+      ts_idx_(ts_idx) {}
+
+DiskTableKeyIterator::~DiskTableKeyIterator() {
+    delete it_;
+    db_->ReleaseSnapshot(snapshot_);
+}
+
+void DiskTableKeyIterator::SeekToFirst() {
+    it_->SeekToFirst();
+    uint32_t cur_ts_idx = UINT32_MAX;
+    ParseKeyAndTs(has_ts_idx_, it_->key(), pk_, ts_, cur_ts_idx);
+}
+
+void DiskTableKeyIterator::NextPK() {
+    std::string last_pk = pk_;
+    std::string combine_key;
+    if (has_ts_idx_) {
+        combine_key = CombineKeyTs(last_pk, 0, ts_idx_);
+    } else {
+        combine_key = CombineKeyTs(last_pk, 0);
+    }
+    it_->Seek(rocksdb::Slice(combine_key));
+    while (it_->Valid()) {
+        uint32_t cur_ts_idx = UINT32_MAX;
+        ParseKeyAndTs(has_ts_idx_, it_->key(), pk_, ts_, cur_ts_idx);
+        if (pk_ != last_pk) {
+            if (has_ts_idx_ && (cur_ts_idx != ts_idx_)) {
+                it_->Next();
+                continue;
+            }
+            // Like KeyIterator in MemTable, maybe no need to check IsExpired here
+            // if (!IsExpired()) {
+            //     return;
+            // }
+            break;
+        } else {
+            it_->Next();
+        }
+    }
+}
+
+void DiskTableKeyIterator::Next() { NextPK(); }
+
+void DiskTableKeyIterator::Seek(const std::string& pk) {
+    std::string combine;
+    uint64_t tmp_ts = UINT64_MAX;
+    if (has_ts_idx_) {
+        combine = CombineKeyTs(pk, tmp_ts, ts_idx_);
+    } else {
+        combine = CombineKeyTs(pk, tmp_ts);
+    }
+    it_->Seek(rocksdb::Slice(combine));
+    for (; it_->Valid(); it_->Next()) {
+        uint32_t cur_ts_idx = UINT32_MAX;
+        ParseKeyAndTs(has_ts_idx_, it_->key(), pk_, ts_, cur_ts_idx);
+        if (pk_ == pk) {
+            if (has_ts_idx_ && (cur_ts_idx != ts_idx_)) {
+                continue;
+            }
+        }
+        break;
+    }
+}
+
+bool DiskTableKeyIterator::Valid() {
+    return it_->Valid();
+}
+
+const hybridse::codec::Row DiskTableKeyIterator::GetKey() {
+    hybridse::codec::Row row(
+        ::hybridse::base::RefCountedSlice::Create(pk_.c_str(), pk_.size()));
+    return row;
+}
+
+std::unique_ptr<::hybridse::vm::RowIterator> DiskTableKeyIterator::GetValue() {
+    std::unique_ptr<DiskTableRowIterator> wit(new DiskTableRowIterator(db_, it_, snapshot_, ttl_type_, expire_time_,
```

Comment:
The `DiskTableRowIterator` has to be valid even after `DiskTableKeyIterator` was destroyed. So cannot pass `it_` directly and share the `it_` and `snapshot_`.

Have to create a new `it` and `snapshot` for each `DiskTableRowIterator`